relayer/chain/ethereum/header_cache_state.go

```diff
@@ -4,13 +4,100 @@
 package ethereum
 
 import (
+	"bytes"
+	"context"
 	"fmt"
+	"math"
 	"sync"
 
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	gethTrie "github.com/ethereum/go-ethereum/trie"
 	"github.com/tranvictor/ethashproof"
 	"golang.org/x/sync/errgroup"
 )
 
+type BlockLoader interface {
+	GetBlock(ctx context.Context, hash gethCommon.Hash) (*gethTypes.Block, error)
+	GetAllReceipts(ctx context.Context, block *gethTypes.Block) (gethTypes.Receipts, error)
+}
+
+type DefaultBlockLoader struct {
+	conn *Connection
+}
+
+func (d *DefaultBlockLoader) GetBlock(ctx context.Context, hash gethCommon.Hash) (*gethTypes.Block, error) {
+	return d.conn.client.BlockByHash(ctx, hash)
+}
+
+func (d *DefaultBlockLoader) GetAllReceipts(ctx context.Context, block *gethTypes.Block) (gethTypes.Receipts, error) {
+	return GetAllReceipts(ctx, d.conn, block)
+}
+
+// Keeps the blocks and receipts for the latest block heights / numbers
+// in memory (up to `capacity` block numbers)
+type BlockCache struct {
+	capacity       int
+	hashesByNumber map[uint64][]string
+	blocks         map[string]*gethTypes.Block
+	receiptTries   map[string]*gethTrie.Trie
+}
+
+func NewBlockCache(capacity int) *BlockCache {
+	return &BlockCache{
+		capacity:       capacity,
+		hashesByNumber: make(map[uint64][]string, capacity),
+		blocks:         make(map[string]*gethTypes.Block, capacity),
+		receiptTries:   make(map[string]*gethTrie.Trie, capacity),
+	}
+}
+
+func (bc *BlockCache) Insert(block *gethTypes.Block, receiptTrie *gethTrie.Trie) {
+	hash := block.Hash().Hex()
+	_, exists := bc.blocks[hash]
+	if exists {
+		return
+	}
+
+	number := block.Number().Uint64()
+	hashesAtNumber, numberExists := bc.hashesByNumber[number]
+	// Remove oldest blocks if we've reached capacity
+	if !numberExists && len(bc.hashesByNumber) == bc.capacity {
```

Comment:
Why only prune if there is no other block at that height?

Can we only skip pruning if the block to be inserted is at the lowest block number?

Reply:
> Why only prune if there is no other block at that height?

If there's no other block at that height, we need to create a new height slot. So, before doing that, we remove the oldest height slot here. `BlockCache.capacity` refers to the max number of heights we'll store, not the max number of blocks (which is a bit confusing, but simplifies the logic).

Reply:
Does my answer resolve your questions?